In [1]:
import pandas as pd
import numpy as np

In [2]:
companies=pd.read_csv("C:\\UpGrad\\companies.txt",delimiter="\t",encoding='ISO-8859-1')
rounds2=pd.read_csv("C:\\UpGrad\\rounds2.csv",encoding='ISO-8859-1')

####   utf 8 is the default encoding but Since the files are encoded with different code So we use encoding ISO-8859-1 to decode it.

- Below some lines of code is to get insight of dataframes.

In [3]:
print(companies.shape)
print(rounds2.shape)

(66368, 10)
(114949, 6)


In [4]:
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB


In [5]:
rounds2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB


In [6]:
(companies.isnull().sum()/len(companies))*100

permalink         0.000000
name              0.001507
homepage_url      7.621143
category_list     4.743250
status            0.000000
country_code     10.483968
state_code       12.878194
region           12.099204
city             12.096191
founded_at       22.934245
dtype: float64

<ul> <li><b>We see that companies DF has higher null values for country_code,state_code,region,city,founded_at,homepage_url</b></li>
<li><b>Most of the columns are of no use like state_code,region,city,founded_at,homepage_url as we'll be doing country wise analysis.We'll drop in the next few steps.</b></li></ul>

In [7]:
companies.describe()

permalink   name                homepage_url category_list  \
count                  66368  66367                       61310         63220   
unique                 66368  66102                       61191         27296   
top     /Organization/Colixo  Roost  http://www.askforoffer.com      Software   
freq                       1      4                           5          3995   

           status country_code state_code       region           city  \
count       66368        59410      57821        58338          58340   
unique          4          137        311         1092           5111   
top     operating          USA         CA  SF Bay Area  San Francisco   
freq        53034        37601      12900         8804           3526   

        founded_at  
count        51147  
unique        3978  
top     01-01-2012  
freq          2730

In [8]:
(companies.isnull().sum()/len(companies))*100

permalink         0.000000
name              0.001507
homepage_url      7.621143
category_list     4.743250
status            0.000000
country_code     10.483968
state_code       12.878194
region           12.099204
city             12.096191
founded_at       22.934245
dtype: float64

In [9]:
(rounds2.isnull().sum()/len(rounds2))*100

company_permalink           0.000000
funding_round_permalink     0.000000
funding_round_type          0.000000
funding_round_code         72.909725
funded_at                   0.000000
raised_amount_usd          17.390321
dtype: float64

- #### rounds2 has higher null value for funding_round_code and raised_amount_usd.
- #### we can drop funding_round_code column as it has higher null value and we are not going to do any analysis on that but we'll not drop raised_amount_usd column as we need to do analysis on amount raised.
- #### we'll drop the column in upcoming steps.

In [10]:
companies

permalink  \
0                                    /Organization/-Fame   
1                                 /Organization/-Qounter   
2                    /Organization/-The-One-Of-Them-Inc-   
3                                  /Organization/0-6-Com   
4                         /Organization/004-Technologies   
5                       /Organization/01Games-Technology   
6                    /Organization/0Ndine-Biomedical-Inc   
7                                   /Organization/0Xdata   
8                                        /Organization/1   
9                              /Organization/1-2-3-Listo   
10                                 /Organization/1-4-All   
11                        /Organization/1-618-Technology   
12                           /Organization/1-800-Dentist   
13                           /Organization/1-800-Doctors   
14              /Organization/1-800-Publicrelations-Inc-   
15                            /Organization/1-Mainstream   
16                                 /Organization/1-Of-99   
17                             /Organization/10-20-Media   
18                         /Organization/10-Minutes-With   
19                              /Organization/1000-Corks   
20                            /Organization/1000-Markets   
21                                 /Organization/1000Chi   
22                               /Organization/1000Lookz   
23                            /Organization/1000Memories   
24                         /Organization/1000Museums-Com   
25                              /Organization/1001-Menus   
26                      /Organization/1001Pharmacies-Com   
27                                 /Organization/1006-Tv   
28                                /Organization/100Du-Tv   
29                                /Organization/100E-Com   
...                                                  ...   
66338                                /Organization/Zykis   
66339                       /Organization/Zylie-The-Bear   
66340                       /Organization/Zylun-Staffing   
66341                       /Organization/Zyme-Solutions   
66342                             /Organization/Zymergen   
66343                              /Organization/Zymetis   
66344                            /Organization/Zymeworks   
66345                             /Organization/Zymochem   
66346                                /Organization/Zympi   
66347                            /Organization/Zymr-Inc-   
66348                                /Organization/Zyncd   
66349                               /Organization/Zyncro   
66350              /Organization/Zynerba-Pharmaceuticals   
66351            /Organization/Zynergy-Projects-Services   
66352                                /Organization/Zynga   
66353                             /Organization/Zyngenia   
66354                              /Organization/Zynstra   
66355                           /Organization/Zyomyx-Inc   
66356                                 /Organization/Zype   
66357                               /Organization/Zypsee   
66358                     /Organization/Zyraz-Technology   
66359                                /Organization/Zyrra   
66360                               /Organization/Zystor   
66361                           /Organization/Zytoprotec   
66362                                /Organization/Zzish   
66363  /Organization/Zznode-Science-And-Technology-Co...   
66364                          /Organization/Zzzzapp-Com   
66365                               /Organization/ÁEron   
66366                             /Organization/ÔAsys-2   
66367  /Organization/Ä°Novatiff-Reklam-Ve-Tanä±Tä±M-H...   

                                                name  \
0                                              #fame   
1                                           :Qounter   
2                             (THE) ONE of THEM,Inc.   
3                                            0-6.com   
4                                   004 Techno

In [11]:
rounds2

company_permalink  \
0                                     /organization/-fame   
1                                  /ORGANIZATION/-QOUNTER   
2                                  /organization/-qounter   
3                     /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                                   /organization/0-6-com   
5                          /ORGANIZATION/004-TECHNOLOGIES   
6                        /organization/01games-technology   
7                     /ORGANIZATION/0NDINE-BIOMEDICAL-INC   
8                     /organization/0ndine-biomedical-inc   
9                                    /ORGANIZATION/0XDATA   
10                                   /organization/0xdata   
11                                   /ORGANIZATION/0XDATA   
12                                   /organization/0xdata   
13                                        /ORGANIZATION/1   
14                                        /organization/1   
15                                        /ORGANIZATION/1   
16                              /organization/1-2-3-listo   
17                                  /ORGANIZATION/1-4-ALL   
18                         /organization/1-618-technology   
19                            /ORGANIZATION/1-800-DENTIST   
20                            /organization/1-800-doctors   
21               /ORGANIZATION/1-800-PUBLICRELATIONS-INC-   
22                             /organization/1-mainstream   
23                                  /ORGANIZATION/1-OF-99   
24                              /organization/10-20-media   
25                              /ORGANIZATION/10-20-MEDIA   
26                              /organization/10-20-media   
27                              /ORGANIZATION/10-20-MEDIA   
28                          /organization/10-minutes-with   
29                          /ORGANIZATION/10-MINUTES-WITH   
...                                                   ...   
114919                              /ORGANIZATION/ZYNSTRA   
114920                           /organization/zyomyx-inc   
114921                           /ORGANIZATION/ZYOMYX-INC   
114922                           /organization/zyomyx-inc   
114923                           /ORGANIZATION/ZYOMYX-INC   
114924                                 /organization/zype   
114925                                 /ORGANIZATION/ZYPE   
114926                               /organization/zypsee   
114927                               /ORGANIZATION/ZYPSEE   
114928                     /organization/zyraz-technology   
114929                     /ORGANIZATION/ZYRAZ-TECHNOLOGY   
114930                     /organization/zyraz-technology   
114931                     /ORGANIZATION/ZYRAZ-TECHNOLOGY   
114932                                /organization/zyrra   
114933                                /ORGANIZATION/ZYRRA   
114934                                /organization/zyrra   
114935                                /ORGANIZATION/ZYRRA   
114936                               /organization/zystor   
114937                           /ORGANIZATION/ZYTOPROTEC   
114938                                /organization/zzish   
114939                                /ORGANIZATION/ZZISH   
114940                                /organization/zzish   
114941  /ORGANIZATION/ZZNODE-SCIENCE-AND-TECHNOLOGY-CO...   
114942                          /organization/zzzzapp-com   
114943                          /ORGANIZATION/ZZZZAPP-COM   
114944                          /organization/zzzzapp-com   
114945                          /ORGANIZATION/ZZZZAPP-COM   
114946                               /organization/ãeron   
114947                             /ORGANIZATION/ÔASYS-2   
114948  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...   

                                funding_round_permalink   funding_round_type  \
0       /funding-round/9a01d05418af9f794eebff7ace91f638              venture   
1       /funding-round/22dacff496eb7acb2b901dec1dfe5633              venture   
2       /funding-

- ##### when we see the permalink in companies and company_permalink in rounds2 both have no null values and since name column has some null values it can not be used for checking unique comapnies.
- ##### Even to merge the both dataframes we'll be using this as this is the common column for both dataframes.
- ##### Now if we see the column values they are encoded,So we need to decode them bring them to common format.
- ##### Since the links are in different case we bring them to same upper or lower case,this will help in both merging and finding unique values

In [12]:
companies['permalink']=companies[['permalink']].applymap(lambda x: x.encode('utf-8').decode('ascii', 'ignore')).applymap(lambda x:x.lower())

In [13]:
len(np.unique(companies.permalink))

66368

In [14]:
rounds2['company_permalink']=rounds2[['company_permalink']].applymap(lambda x: x.encode('utf-8').decode('ascii', 'ignore')).applymap(lambda x:x.lower())

In [15]:
len(np.unique(rounds2['company_permalink']))

66368

- From the above outputs we can see that 66368 unique comapanies are present in both rounds2 and companies dataframe

- ##### lengths are same but it does not mean the values are same.

In [16]:
sum(np.unique(rounds2[['company_permalink']])!=np.unique(companies[['permalink']]))

0

- ##### When we compare the values and take the sum for not equal values it's 0 from the above output which shows that same set of companies are present in both the tables.

- Finally we'll merge both dataframes to form single master dataframe which has both fundings and details of companies.
- we merge on permalink in both dataframes as the values are same and not null

In [17]:
master_frame=rounds2.merge(companies,left_on='company_permalink',right_on='permalink')

In [18]:
master_frame.shape

(114949, 16)

- their are 114949 rows and 16 columns present.

### Checkpoint 1 Complete

In [19]:
master_frame.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 114949 entries, 0 to 114948
Data columns (total 16 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
permalink                  114949 non-null object
name                       114948 non-null object
homepage_url               108815 non-null object
category_list              111539 non-null object
status                     114949 non-null object
country_code               106271 non-null object
state_code                 104003 non-null object
region                     104782 non-null object
city                       104785 non-null object
founded_at                 94428 non-null object
dtypes: float64(1), object(15)
memory usage: 14.9+ MB


In [20]:
(master_frame.isnull().sum()/len(master_frame))*100

company_permalink           0.000000
funding_round_permalink     0.000000
funding_round_type          0.000000
funding_round_code         72.909725
funded_at                   0.000000
raised_amount_usd          17.390321
permalink                   0.000000
name                        0.000870
homepage_url                5.336280
category_list               2.966533
status                      0.000000
country_code                7.549435
state_code                  9.522484
region                      8.844792
city                        8.842182
founded_at                 17.852265
dtype: float64

- Now we have the single dataframe on which we can do the cleaning and analysis
- So we finally drop the columns which are of no use for analysis and columns with high null values.
- Since permalink is appearing twice due to merging we drop one of the column,raised amount is not dropped as we need that column for analysis we fill the null values in next steps instead of dropping it.

In [21]:
master_frame.drop(['funding_round_permalink','funding_round_code','permalink','homepage_url','state_code','region','city','founded_at'],axis=1,inplace=True)

In [22]:
master_frame.head()

company_permalink funding_round_type   funded_at  \
0                  /organization/-fame            venture  05-01-2015   
1               /organization/-qounter            venture  14-10-2014   
2               /organization/-qounter               seed  01-03-2014   
3  /organization/-the-one-of-them-inc-            venture  30-01-2014   
4                /organization/0-6-com            venture  19-03-2008   

   raised_amount_usd                    name  \
0         10000000.0                   #fame   
1                NaN                :Qounter   
2           700000.0                :Qounter   
3          3406878.0  (THE) ONE of THEM,Inc.   
4          2000000.0                 0-6.com   

                                       category_list     status country_code  
0                                              Media  operating          IND  
1  Application Platforms|Real Time|Social Network...  operating          USA  
2  Application Platforms|Real Time|Social Network...  operating          USA  
3                                  Apps|Games|Mobile  operating          NaN  
4                                        Curated Web  operating          CHN

In [23]:
(master_frame.isnull().sum()/len(master_frame))*100

company_permalink      0.000000
funding_round_type     0.000000
funded_at              0.000000
raised_amount_usd     17.390321
name                   0.000870
category_list          2.966533
status                 0.000000
country_code           7.549435
dtype: float64

In [24]:
np.unique(master_frame['funding_round_type'])

array(['angel', 'convertible_note', 'debt_financing',
       'equity_crowdfunding', 'grant', 'non_equity_assistance',
       'post_ipo_debt', 'post_ipo_equity', 'private_equity',
       'product_crowdfunding', 'secondary_market', 'seed', 'undisclosed',
       'venture'], dtype=object)

- We drop duplicate rows as it does not make sense to keep the same data twice.

In [25]:
master_frame.drop_duplicates(inplace=True)

In [26]:
master_frame[master_frame.isnull().sum(axis=1)>2].head()

company_permalink funding_round_type   funded_at  \
201             /organization/2-ladoshki               seed  24-12-2013   
275               /organization/24med-sp     private_equity  01-01-2008   
291                 /organization/29west        undisclosed  15-03-2007   
366  /organization/30secondstofly-claire        undisclosed  30-09-2015   
439               /organization/3d-forms        undisclosed  16-09-2013   

     raised_amount_usd                     name category_list     status  \
201                NaN               2 Ladoshki           NaN  operating   
275                NaN                 24med Sp           NaN     closed   
291                NaN                   29West           NaN   acquired   
366                NaN  30SecondsToFly - Claire           NaN     closed   
439                NaN                 3D Forms           NaN  operating   

    country_code  
201          NaN  
275          NaN  
291          NaN  
366          NaN  
439          NaN

- We need country,category and amount raised to do the analysis on country for particular sector how much amount is present,if any of the rows contains more than 2 null values as only raised_amount_usd,category_list and country_code have nulls,Such a row will be of no use.So we drop those rows

In [27]:
master_frame.drop(master_frame.index[master_frame.isnull().sum(axis=1)>2],inplace=True)

In [28]:
(master_frame.isnull().sum()/len(master_frame))*100

company_permalink      0.000000
funding_round_type     0.000000
funded_at              0.000000
raised_amount_usd     16.498349
name                   0.000880
category_list          1.937046
status                 0.000000
country_code           6.570988
dtype: float64

- ##### We can see from above output country_code,category_list and raised_amount_usd have null values .
- ##### we drop rows having country_code,category_list null instead of making some assumption which may hamper our analysis.
- ##### we do not drop rows having raised_amount_usd null as the percentage of rows is more which will lead to loss of more data.
- ##### Since we need to do analysis based on funding round type which are different type,So we cannot substitute common amount for all null values in column raised_amount_usd.We need to take the mean based on the funding_round_type and the substitute the value.
- ##### Mean because it will not change the average value for the funding_round_type any other value will increase/decrease the value.

In [29]:
master_frame.shape[0]/114949

0.9880468729610523

- Like we discussed above we drop the rows.

In [30]:
master_frame.drop(master_frame.index[master_frame.country_code.isnull()],inplace=True)

In [31]:
master_frame.drop(master_frame.index[master_frame.category_list.isnull()],inplace=True)

In [32]:
master_frame.shape[0]/114949

0.9080200784695822

- We still have 90.8% of data compared to original data.

In [33]:
(master_frame.isnull().sum()/len(master_frame))*100

company_permalink      0.000000
funding_round_type     0.000000
funded_at              0.000000
raised_amount_usd     15.294704
name                   0.000958
category_list          0.000000
status                 0.000000
country_code           0.000000
dtype: float64

In [34]:
ftypes=np.unique(master_frame['funding_round_type'])

- Now we substitute the mean values for raised_amount_usd per funding_round_type in place of null values. 

In [35]:
for i in ftypes:
    master_frame.loc[(master_frame['funding_round_type']==i) & (master_frame['raised_amount_usd'].isnull()),'raised_amount_usd']=master_frame.loc[(master_frame['funding_round_type']==i),'raised_amount_usd'].mean()

- ##### since they want to invest the value in millions we divide the mean by 1 million.
- ##### They want to invest between 5-15 million.

In [36]:
master_frame.loc[master_frame['funding_round_type']=='angel','raised_amount_usd'].mean()/1000000

0.972806151025062

- Angel has only 0.97million investment on an average which is too less

In [37]:
master_frame.loc[master_frame['funding_round_type']=='venture','raised_amount_usd'].mean()/1000000

11.729266434695855

- Venture has 11.73 million on an average which suits our criteria.

In [38]:
master_frame.loc[master_frame['funding_round_type']=='seed','raised_amount_usd'].mean()/1000000

0.7478570231664283

- Seed has only 0.75million investment on an average which is too less

In [39]:
master_frame.loc[master_frame['funding_round_type']=='private_equity','raised_amount_usd'].mean()/1000000

73.93848627527439

- Private Equity has only 73.94million investment on an average which is too High

- ##### We choose Venture as the investment type for spark funds

### Checkpoint 2 complete

In [40]:
(master_frame.isnull().sum()/len(master_frame))*100

company_permalink     0.000000
funding_round_type    0.000000
funded_at             0.000000
raised_amount_usd     0.000000
name                  0.000958
category_list         0.000000
status                0.000000
country_code          0.000000
dtype: float64

- From the above output we can see that name has some null values which we can ignore instead of adding some random value as we are not using this column for analysis,even if we want we can drop too but in checkpoint 5 we need to know the company we keep this column withot modifying.

#### Now we want to know the top 9 countries having highest investment in all the sectors.
- ##### We create pivot table to determine it

In [41]:
all_countries_invest=pd.pivot_table(master_frame[master_frame['funding_round_type']=='venture'],index=['country_code','funding_round_type'],values=['raised_amount_usd'],aggfunc=np.sum)

- We extract top 9 countries

In [42]:
top9=all_countries_invest.sort_values(by='raised_amount_usd',ascending=False).head(9)

In [43]:
top9.reset_index(level=['country_code','funding_round_type'],inplace=True)

In [44]:
top9.drop(['funding_round_type'],axis=1,inplace=True)

In [45]:
top9

country_code  raised_amount_usd
0          USA       4.443780e+11
1          CHN       4.519074e+10
2          GBR       2.270017e+10
3          IND       1.617338e+10
4          CAN       1.046748e+10
5          DEU       8.863902e+09
6          ISR       8.151299e+09
7          FRA       7.918878e+09
8          JPN       3.625089e+09

#### Above list shows the top 9 countries with highest investment in all sectors.

- ##### We want to know the top 3 english speaking countries as the spark fund wants to invest only in english speaking countries.

In [46]:
countrywithcodes=pd.read_csv('C:\\UpGrad\\EngCountry.csv')

- Above mentioned file contains all the english speaking countries and their codes.

In [47]:
countrywithcodes

Country CountryCode
0                Antigua and Barbuda         ATG
1                          Australia         AUS
2                           Barbados         BRB
3                             Belize         BLZ
4                           Botswana         BWA
5                           Cameroon         CMR
6                             Canada         CAN
7                           Dominica         DMA
8                            Eritrea         ERI
9                           Ethiopia         ETH
10                              Fiji         FJI
11                             Ghana         GHA
12                           Grenada         GRD
13                            Guyana         GUY
14                             India         IND
15                           Ireland         IRL
16                           Jamaica         JAM
17                             Kenya         KEN
18                          Kiribati         KIR
19                           Lesotho         LSO
20                           Liberia         LBR
21                            Malawi         MWI
22                             Malta         MLT
23                  Marshall Islands         MHL
24                         Mauritius         MUS
25                           Namibia         NAM
26                             Nauru         NRU
27                       New Zealand         NZL
28                           Nigeria         NGA
29                          Pakistan         PAK
30                             Palau         PLW
31                  Papua New Guinea         PNG
32                       Philippines         PHL
33                            Rwanda         RWA
34             Saint Kitts and Nevis         KNA
35                       Saint Lucia         LCA
36  Saint Vincent and the Grenadines         VCT
37                             Samoa         WSM
38                        Seychelles         SYC
39                      Sierra Leone         SLE
40                         Singapore         SGP
41                   Solomon Islands         SLB
42                      South Africa         ZAF
43                       South Sudan         SSD
44                             Sudan         SDN
45                         Swaziland         SWZ
46                          Tanzania         TZA
47                             Tonga         TON
48               Trinidad and Tobago         TTO
49                            Tuvalu         TUV
50                            Uganda         UGA
51                    United Kingdom         GBR
52                     United States         USA
53                           Vanuatu         VUT
54                            Zambia         ZMB
55                          Zimbabwe         ZWE

In [48]:
top9.loc[top9['country_code'].map(lambda x:x in list(countrywithcodes['CountryCode']))].head(3)

country_code  raised_amount_usd
0          USA       4.443780e+11
2          GBR       2.270017e+10
3          IND       1.617338e+10

- ##### From the above output we can see that USA,GBR and IND are the winners.They can make investment in these countries.

### Checkpoint 3 complete

- We know the country and investment type now we need to decide the sector.We have multiple categories,So we use the mapping file to map them to one main sector out of 8.

In [49]:
mapping=pd.read_csv('C:\\UpGrad\\mapping.csv')

In [50]:
import re

In [51]:
def removezeros(data):
    ret_val=[]
    if data==data:
        for i in data.split(' '):
            if re.match('[a-zA-Z]*0[a-zA-Z]*',i)!=None:
                ret_val.append(i.replace('0','na'))
            else:
                ret_val.append(i)
    return ' '.join(ret_val).title()

- Above function will remove the `0` present insted of `na` in category list of mapping file as data cleaning process and put `na` in place `0` using regex.
- Since we need to merge them,So we bring to comman case in this case title(camelcase)

In [52]:
mapping['category_list']=mapping['category_list'].map(removezeros)

In [53]:
master_frame['primary_sector']=master_frame['category_list'].map(lambda x:x.split('|')[0].title())

- Their are multiple categories so we take the first category as primary category by doing the above operation.

In [54]:
master_frame.head()

company_permalink funding_round_type   funded_at  \
0             /organization/-fame            venture  05-01-2015   
1          /organization/-qounter            venture  14-10-2014   
2          /organization/-qounter               seed  01-03-2014   
4           /organization/0-6-com            venture  19-03-2008   
5  /organization/004-technologies            venture  24-07-2014   

   raised_amount_usd              name  \
0       1.000000e+07             #fame   
1       1.172927e+07          :Qounter   
2       7.000000e+05          :Qounter   
4       2.000000e+06           0-6.com   
5       1.172927e+07  004 Technologies   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2  Application Platforms|Real Time|Social Network...  operating          USA   
4                                        Curated Web  operating          CHN   
5                                           Software  operating          USA   

          primary_sector  
0                  Media  
1  Application Platforms  
2  Application Platforms  
4            Curated Web  
5               Software

In [55]:
final_df=master_frame.merge(mapping,left_on='primary_sector',right_on='category_list',how='left')

- Merging the dataframes to form the dataframe with sectors mapped.

In [56]:
final_df.head()

company_permalink funding_round_type   funded_at  \
0             /organization/-fame            venture  05-01-2015   
1          /organization/-qounter            venture  14-10-2014   
2          /organization/-qounter               seed  01-03-2014   
3           /organization/0-6-com            venture  19-03-2008   
4  /organization/004-technologies            venture  24-07-2014   

   raised_amount_usd              name  \
0       1.000000e+07             #fame   
1       1.172927e+07          :Qounter   
2       7.000000e+05          :Qounter   
3       2.000000e+06           0-6.com   
4       1.172927e+07  004 Technologies   

                                     category_list_x     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2  Application Platforms|Real Time|Social Network...  operating          USA   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

          primary_sector        category_list_y  Automotive & Sports  Blanks  \
0                  Media                  Media                  0.0     0.0   
1  Application Platforms  Application Platforms                  0.0     0.0   
2  Application Platforms  Application Platforms                  0.0     0.0   
3            Curated Web            Curated Web                  0.0     0.0   
4               Software               Software                  0.0     0.0   

   Cleantech / Semiconductors  Entertainment  Health  Manufacturing  \
0                         0.0            1.0     0.0            0.0   
1                         0.0            0.0     0.0            0.0   
2                         0.0            0.0     0.0            0.0   
3                         0.0            0.0     0.0            0.0   
4                         0.0            0.0     0.0            0.0   

   News, Search and Messaging  Others  Social, Finance, Analytics, Advertising  
0                         0.0     0.0                                      0.0  
1                         1.0     0.0                                      0.0  
2                         1.0     0.0                                      0.0  
3                         1.0     0.0                                      0.0  
4                         0.0     1.0                                      0.0

In [57]:
final_df.drop(['funded_at','status'],axis=1,inplace=True)

- Dropping useless columns for convinience of analysis.

In [58]:
final_df[final_df['category_list_y'].isnull()].head()

company_permalink  \
2372                             /organization/adora-inc   
5805   /organization/apollo-commercial-real-estate-fi...   
7447                        /organization/asia-translate   
10273                 /organization/bellwether-institute   
10274                 /organization/bellwether-institute   

          funding_round_type  raised_amount_usd  \
2372                    seed       6.000000e+04   
5805          debt_financing       5.000000e+07   
7447                 venture       2.000000e+04   
10273                   seed       7.478570e+05   
10274  non_equity_assistance       4.807534e+05   

                                        name  \
2372                              Adora Inc.   
5805   Apollo Commercial Real Estate Finance   
7447                          Asia Translate   
10273          Bellwether Research Institute   
10274          Bellwether Research Institute   

                                         category_list_x country_code  \
2372                             Cause Marketing|Charity          USA   
5805                               Real Estate Investors          USA   
7447                        English-Speaking|Translation          SGP   
10273  Deep Information Technology|Emerging Markets|E...          USA   
10274  Deep Information Technology|Emerging Markets|E...          USA   

                    primary_sector category_list_y  Automotive & Sports  \
2372               Cause Marketing             NaN                  NaN   
5805         Real Estate Investors             NaN                  NaN   
7447              English-Speaking             NaN                  NaN   
10273  Deep Information Technology             NaN                  NaN   
10274  Deep Information Technology             NaN                  NaN   

       Blanks  Cleantech / Semiconductors  Entertainment  Health  \
2372      NaN                         NaN            NaN     NaN   
5805      NaN                         NaN            NaN     NaN   
7447      NaN                         NaN            NaN     NaN   
10273     NaN                         NaN            NaN     NaN   
10274     NaN                         NaN            NaN     NaN   

       Manufacturing  News, Search and Messaging  Others  \
2372             NaN                         NaN     NaN   
5805             NaN                         NaN     NaN   
7447             NaN                         NaN     NaN   
10273            NaN                         NaN     NaN   
10274            NaN                         NaN     NaN   

       Social, Finance, Analytics, Advertising  
2372                                       NaN  
5805                                       NaN  
7447                                       NaN  
10273                                      NaN  
10274                                      NaN

- ##### From the above output we can make out that their are some primary sectors which are not present in the mapping file.
- ##### So we map them to others sector as they are also one of the important sectors,So we cannot remove them from analysis.

In [59]:
final_df.loc[final_df['category_list_y'].isnull(),['Automotive & Sports','Blanks','Cleantech / Semiconductors','Entertainment','Health','Manufacturing','News, Search and Messaging','Others','Social, Finance, Analytics, Advertising']]=[0,0,0,0,0,0,0,1,0]

- ##### We need to create common main sector instead of maintaining these many sectors in seperate columns.

In [60]:
def test(row):
    if row['Automotive & Sports']==1.0:
        return 'Automotive & Sports'
    elif row['Cleantech / Semiconductors']==1.0:
        return 'Cleantech / Semiconductors'
    elif row['Entertainment']==1.0:
        return 'Entertainment'
    elif row['Health']==1.0:
        return 'Health'
    elif row['Manufacturing']==1.0:
        return 'Manufacturing'
    elif row['News, Search and Messaging']==1.0:
        return 'News, Search and Messaging'
    elif row['Others']==1.0:
        return 'Others'
    else:
        return 'Social, Finance, Analytics, Advertising'
    

- ##### Above function will determine the main sector along the row which will be added to Main Sector column.

In [61]:
final_df['Main Sector']=final_df.apply(test,axis=1)

#### Checkpoint 4 Complete

In [62]:
final_df.head()

company_permalink funding_round_type  raised_amount_usd  \
0             /organization/-fame            venture       1.000000e+07   
1          /organization/-qounter            venture       1.172927e+07   
2          /organization/-qounter               seed       7.000000e+05   
3           /organization/0-6-com            venture       2.000000e+06   
4  /organization/004-technologies            venture       1.172927e+07   

               name                                    category_list_x  \
0             #fame                                              Media   
1          :Qounter  Application Platforms|Real Time|Social Network...   
2          :Qounter  Application Platforms|Real Time|Social Network...   
3           0-6.com                                        Curated Web   
4  004 Technologies                                           Software   

  country_code         primary_sector        category_list_y  \
0          IND                  Media                  Media   
1          USA  Application Platforms  Application Platforms   
2          USA  Application Platforms  Application Platforms   
3          CHN            Curated Web            Curated Web   
4          USA               Software               Software   

   Automotive & Sports  Blanks  Cleantech / Semiconductors  Entertainment  \
0                  0.0     0.0                         0.0            1.0   
1                  0.0     0.0                         0.0            0.0   
2                  0.0     0.0                         0.0            0.0   
3                  0.0     0.0                         0.0            0.0   
4                  0.0     0.0                         0.0            0.0   

   Health  Manufacturing  News, Search and Messaging  Others  \
0     0.0            0.0                         0.0     0.0   
1     0.0            0.0                         1.0     0.0   
2     0.0            0.0                         1.0     0.0   
3     0.0            0.0                         1.0     0.0   
4     0.0            0.0                         0.0     1.0   

   Social, Finance, Analytics, Advertising                 Main Sector  
0                                      0.0               Entertainment  
1                                      0.0  News, Search and Messaging  
2                                      0.0  News, Search and Messaging  
3                                      0.0  News, Search and Messaging  
4                                      0.0                      Others

In [63]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 104376 entries, 0 to 104375
Data columns (total 18 columns):
company_permalink                          104376 non-null object
funding_round_type                         104376 non-null object
raised_amount_usd                          104376 non-null float64
name                                       104375 non-null object
category_list_x                            104376 non-null object
country_code                               104376 non-null object
primary_sector                             104376 non-null object
category_list_y                            104303 non-null object
Automotive & Sports                        104376 non-null float64
Blanks                                     104376 non-null float64
Cleantech / Semiconductors                 104376 non-null float64
Entertainment                              104376 non-null float64
Health                                     104376 non-null float64
Manufacturing            

In [64]:
final_df.isnull().sum()/len(final_df)

company_permalink                          0.000000
funding_round_type                         0.000000
raised_amount_usd                          0.000000
name                                       0.000010
category_list_x                            0.000000
country_code                               0.000000
primary_sector                             0.000000
category_list_y                            0.000699
Automotive & Sports                        0.000000
Blanks                                     0.000000
Cleantech / Semiconductors                 0.000000
Entertainment                              0.000000
Health                                     0.000000
Manufacturing                              0.000000
News, Search and Messaging                 0.000000
Others                                     0.000000
Social, Finance, Analytics, Advertising    0.000000
Main Sector                                0.000000
dtype: float64

- ##### Now we know the top 3 countries where they want to invest and investment type.We go country wise which sector is the best

- We take out only the relavant columns for the perticular country and venture investment type determined earlier.

In [65]:
D1=final_df.loc[(final_df['country_code']=='USA') & (final_df['funding_round_type']=='venture'),['company_permalink','funding_round_type','raised_amount_usd','name','category_list_x','country_code','primary_sector','Main Sector']]

In [66]:
D1.head()

company_permalink funding_round_type  raised_amount_usd  \
1           /organization/-qounter            venture       1.172927e+07   
4   /organization/004-technologies            venture       1.172927e+07   
9             /organization/0xdata            venture       2.000000e+07   
10            /organization/0xdata            venture       1.700000e+06   
11            /organization/0xdata            venture       8.900000e+06   

                name                                    category_list_x  \
1           :Qounter  Application Platforms|Real Time|Social Network...   
4   004 Technologies                                           Software   
9             H2O.ai                                          Analytics   
10            H2O.ai                                          Analytics   
11            H2O.ai                                          Analytics   

   country_code         primary_sector  \
1           USA  Application Platforms   
4           USA               Software   
9           USA              Analytics   
10          USA              Analytics   
11          USA              Analytics   

                                Main Sector  
1                News, Search and Messaging  
4                                    Others  
9   Social, Finance, Analytics, Advertising  
10  Social, Finance, Analytics, Advertising  
11  Social, Finance, Analytics, Advertising

In [67]:
D2=final_df.loc[(final_df['country_code']=='GBR') & (final_df['funding_round_type']=='venture'),['company_permalink','funding_round_type','raised_amount_usd','name','category_list_x','country_code','primary_sector','Main Sector']]

In [68]:
D3=final_df.loc[(final_df['country_code']=='IND') & (final_df['funding_round_type']=='venture'),['company_permalink','funding_round_type','raised_amount_usd','name','category_list_x','country_code','primary_sector','Main Sector']]

In [69]:
mainsectors=np.unique(D1['Main Sector'])

In [70]:
for i in mainsectors:
    D1.loc[D1['Main Sector']==i,'count_per_sector']=D1.loc[D1['Main Sector']==i,'Main Sector'].count()
    D2.loc[D2['Main Sector']==i,'count_per_sector']=D2.loc[D2['Main Sector']==i,'Main Sector'].count()
    D3.loc[D3['Main Sector']==i,'count_per_sector']=D3.loc[D3['Main Sector']==i,'Main Sector'].count()
    D1.loc[D1['Main Sector']==i,'sum_amount_per_sec']=D1.loc[D1['Main Sector']==i,'raised_amount_usd'].sum()
    D2.loc[D2['Main Sector']==i,'sum_amount_per_sec']=D2.loc[D2['Main Sector']==i,'raised_amount_usd'].sum()
    D3.loc[D3['Main Sector']==i,'sum_amount_per_sec']=D3.loc[D3['Main Sector']==i,'raised_amount_usd'].sum()

- Since we want to add 2 new columns which states count per sector and amount investment per sector above line of code does it

In [71]:
D1.head()

company_permalink funding_round_type  raised_amount_usd  \
1           /organization/-qounter            venture       1.172927e+07   
4   /organization/004-technologies            venture       1.172927e+07   
9             /organization/0xdata            venture       2.000000e+07   
10            /organization/0xdata            venture       1.700000e+06   
11            /organization/0xdata            venture       8.900000e+06   

                name                                    category_list_x  \
1           :Qounter  Application Platforms|Real Time|Social Network...   
4   004 Technologies                                           Software   
9             H2O.ai                                          Analytics   
10            H2O.ai                                          Analytics   
11            H2O.ai                                          Analytics   

   country_code         primary_sector  \
1           USA  Application Platforms   
4           USA               Software   
9           USA              Analytics   
10          USA              Analytics   
11          USA              Analytics   

                                Main Sector  count_per_sector  \
1                News, Search and Messaging            4609.0   
4                                    Others            8772.0   
9   Social, Finance, Analytics, Advertising            7811.0   
10  Social, Finance, Analytics, Advertising            7811.0   
11  Social, Finance, Analytics, Advertising            7811.0   

    sum_amount_per_sec  
1         4.862895e+10  
4         8.832456e+10  
9         9.299952e+10  
10        9.299952e+10  
11        9.299952e+10

In [72]:
print('Total number of investments for D1 :{0}'.format(len(D1)))
print('Total number of investments for D2 :{0}'.format(len(D2)))
print('Total number of investments for D3 :{0}'.format(len(D3)))

Total number of investments for D1 :37996
Total number of investments for D2 :2265
Total number of investments for D3 :982


In [73]:
print('Total amount of investment for D1 :{0}'.format(D1['raised_amount_usd'].sum()))
print('Total amount of investment for D2 :{0}'.format(D2['raised_amount_usd'].sum()))
print('Total amount of investment for D3 :{0}'.format(D3['raised_amount_usd'].sum()))

Total amount of investment for D1 :444377974114.2109
Total amount of investment for D2 :22700168685.371933
Total amount of investment for D3 :16173379146.855473


In [74]:
D1['Main Sector'].astype('category').value_counts()

Others                                     8772
Cleantech / Semiconductors                 8263
Social, Finance, Analytics, Advertising    7811
News, Search and Messaging                 4609
Health                                     3408
Manufacturing                              2642
Entertainment                              1925
Automotive & Sports                         566
Name: Main Sector, dtype: int64

#### From the above output we can see that top 3 sectors for USA count wise are `Others` and `Cleantech / Semiconductors` and `Social, Finance, Analytics, Advertising` .

- #### corresponding Number of invesrment is  `Others=8772` and `Cleantech / Semiconductors=8263` and `Social, Finance, Analytics, Advertising=7811` .

In [75]:
D1[D1['Main Sector']=='Others'].sort_values(by='raised_amount_usd',ascending=False).head(1)

company_permalink funding_round_type  raised_amount_usd  \
83317  /organization/social-finance            venture       1.000000e+09   

       name                                    category_list_x country_code  \
83317  SoFi  Education|Finance|Financial Services|Peer-to-Peer          USA   

      primary_sector Main Sector  count_per_sector  sum_amount_per_sec  
83317      Education      Others            8772.0        8.832456e+10

#### Company that received highest investment for `Others` sector is `SoFi`

In [76]:
D1[D1['Main Sector']=='Cleantech / Semiconductors'].sort_values(by='raised_amount_usd',ascending=False).head(1)

company_permalink funding_round_type  raised_amount_usd  \
34324  /organization/freescale            venture       1.760000e+10   

                          name category_list_x country_code  primary_sector  \
34324  Freescale Semiconductor  Semiconductors          USA  Semiconductors   

                      Main Sector  count_per_sector  sum_amount_per_sec  
34324  Cleantech / Semiconductors            8263.0        1.230855e+11

#### Company that received highest investment for `Cleantech / Semiconductors` sector is `Freescale Semiconductor`

In [77]:
D2['Main Sector'].astype('category').value_counts()

Others                                     581
Social, Finance, Analytics, Advertising    481
Cleantech / Semiconductors                 466
News, Search and Messaging                 265
Entertainment                              160
Manufacturing                              132
Health                                     126
Automotive & Sports                         54
Name: Main Sector, dtype: int64

#### From the above output we can see that top 3 sectors for GBR count wise are `Others`  and `Social, Finance, Analytics, Advertising` and `Cleantech / Semiconductors` .

- #### corresponding Number of invesrment is  `Others=581`  and `Social, Finance, Analytics, Advertising=481` and `Cleantech / Semiconductors=466`.

In [78]:
D2[D2['Main Sector']=='Others'].sort_values(by='raised_amount_usd',ascending=False).head(1)

company_permalink funding_round_type  raised_amount_usd    name  \
64185  /organization/oneweb            venture        500000000.0  OneWeb   

      category_list_x country_code primary_sector Main Sector  \
64185      Technology          GBR     Technology      Others   

       count_per_sector  sum_amount_per_sec  
64185             581.0        5.311462e+09

#### Company that received highest investment for `Others` sector is `OneWeb`

In [79]:
D2[D2['Main Sector']=='Social, Finance, Analytics, Advertising'].sort_values(by='raised_amount_usd',ascending=False).head(1)

company_permalink funding_round_type  raised_amount_usd   name  \
55513  /organization/mereo            venture        119000000.0  Mereo   

      category_list_x country_code primary_sector  \
55513      Consulting          GBR     Consulting   

                                   Main Sector  count_per_sector  \
55513  Social, Finance, Analytics, Advertising             481.0   

       sum_amount_per_sec  
55513        4.477085e+09

#### Company that received highest investment for `Social, Finance, Analytics, Advertising` sector is `Mereo`.

In [80]:
D3['Main Sector'].astype('category').value_counts()

Others                                     332
Social, Finance, Analytics, Advertising    193
News, Search and Messaging                 154
Entertainment                               90
Cleantech / Semiconductors                  63
Manufacturing                               62
Health                                      56
Automotive & Sports                         32
Name: Main Sector, dtype: int64

#### From the above output we can see that top 3 sectors for IND count wise are `Others`  and `Social, Finance, Analytics, Advertising` and `News, Search and Messaging` .

- #### corresponding Number of invesrment is  `Others=332`  and `Social, Finance, Analytics, Advertising=193` and `News, Search and Messaging=154`.

In [81]:
D3[D3['Main Sector']=='Others'].sort_values(by='raised_amount_usd',ascending=False).head(1)

company_permalink funding_round_type  raised_amount_usd      name  \
33255  /organization/flipkart            venture       1.000000e+09  Flipkart   

                  category_list_x country_code primary_sector Main Sector  \
33255  E-Commerce|Online Shopping          IND     E-Commerce      Others   

       count_per_sector  sum_amount_per_sec  
33255             332.0        6.892512e+09

#### Company that received highest investment for `Others` sector is `Flipkart`.

In [82]:
D3[D3['Main Sector']=='Social, Finance, Analytics, Advertising'].sort_values(by='raised_amount_usd',ascending=False).head(1)

company_permalink funding_round_type  raised_amount_usd  \
80832  /organization/shopclues-com            venture        100000000.0   

                name                        category_list_x country_code  \
80832  ShopClues.com  Consumer Internet|E-Commerce|Shopping          IND   

          primary_sector                              Main Sector  \
80832  Consumer Internet  Social, Finance, Analytics, Advertising   

       count_per_sector  sum_amount_per_sec  
80832             193.0        2.118351e+09

#### Company that received highest investment for `Social, Finance, Analytics, Advertising` sector is `ShopClues.com`.

### Since our objective is to  invest in sector where maximum number of people are investing,As per the above analysis `Others` sector tops all the 3 countries.So it is recommended to invest in `Others` sector

### Checkpoint 5 complete

- #### Now we export final dataframe to csv to plot in Tableau

In [83]:
#final_df.to_csv('C:\\UpGrad\\Investment_Case_Study\\finaldata.csv')

### Checkpoint 6 data export complete,Plots in tableau ref the presentation to see the results

- #### Below code is just to see average investments received per main sector.

In [84]:
def getinmillions(data):
    return np.mean(data)/1000000

In [85]:
D1.pivot_table(index=['Main Sector'],values=['raised_amount_usd'],aggfunc=getinmillions).sort_values(by='raised_amount_usd',ascending=False)

raised_amount_usd
Main Sector                                               
Automotive & Sports                              21.815636
Cleantech / Semiconductors                       14.895987
Social, Finance, Analytics, Advertising          11.906224
News, Search and Messaging                       10.550869
Manufacturing                                    10.247229
Others                                           10.068919
Health                                            9.786188
Entertainment                                     9.645323

In [86]:
D2.pivot_table(index=['Main Sector'],values=['raised_amount_usd'],aggfunc=getinmillions).sort_values(by='raised_amount_usd',ascending=False)

raised_amount_usd
Main Sector                                               
News, Search and Messaging                       12.217080
Cleantech / Semiconductors                       11.490971
Automotive & Sports                              11.077248
Social, Finance, Analytics, Advertising           9.307868
Others                                            9.141931
Health                                            9.051500
Entertainment                                     9.025420
Manufacturing                                     8.610423

In [87]:
D3.pivot_table(index=['Main Sector'],values=['raised_amount_usd'],aggfunc=getinmillions).sort_values(by='raised_amount_usd',ascending=False)

raised_amount_usd
Main Sector                                               
Automotive & Sports                              49.002477
Others                                           20.760579
Cleantech / Semiconductors                       14.158858
News, Search and Messaging                       13.123911
Manufacturing                                    13.102659
Health                                           13.010352
Entertainment                                    12.671128
Social, Finance, Analytics, Advertising          10.975913